### PR TITLE
#272 error collection

### PR DESCRIPTION
**Pull request by elikoga**
#272

### CODE REVIEW

File: src/batou/environment.py [hunk 1]
```diff
@@ -454,15 +467,23 @@ def configure(self):
...
                     exceptions.append(
-                        UnknownComponentConfigurationError(root, e, tb)
+                        UnknownComponentConfigurationError.from_context(
+                            root, e, tb
+                        )
                     )
-                    retry.add(root)
+                else:
+                    # configured this component successfully
+                    # go to next component
```
Comment (ctheune) [inline src/batou/environment.py:484]: go to next component is what the "continue" already does. intentional this is:

configured this component successfully, so we won't have to retry it later

File: src/batou/remote_core.py [hunk 2]
```diff
@@ -353,7 +358,12 @@ def setup_deployment(*args):
     os.chdir(deployment_base)
     global deployment
     deployment = Deployment(*args)
-    deployment.load()
+    errors = deployment.load()
+    # pickle errors
```
Comment (ctheune) [inline src/batou/remote_core.py:362]: or just do return pickle.dumps(errors)

File: src/batou/tests/test_dependencies.py [hunk 3]
```diff
@@ -97,32 +100,38 @@ def test_consumer_retrieves_value_from_provider_order2(env):
...
 
 def test_broken_component_logs_real_exception(env):
-    env.add_root("broken", "host")
-    with pytest.raises(UnknownComponentConfigurationError):
-        env.configure()
+    env.add_root("broken", Host("host", env))
+    errors = env.configure()
+    assert len(errors) == 2
+    # could be either one of the two exceptions
+    error_types = (UnknownComponentConfigurationError, NonConvergingWorkingSet)
```
Comment (ctheune) [inline src/batou/tests/test_dependencies.py:132]: This is surprising. I wonder whether NonConvergingWorkingSet should not be reported - in this case. I'm not sure whether they can happen without any other (Non-SilentConfigurationErrors) happening and thus we might want to only report them if they appear alone. Let's make a follow-up ticket for this.

File: src/batou/tests/test_endtoend.py [hunk 4]
```diff
@@ -100,6 +98,9 @@ def test_example_errors_gpg_cannot_decrypt(monkeypatch):
 def test_example_errors_late():
     os.chdir("examples/errors2")
     out, _ = cmd("./batou deploy errors", acceptable_returncodes=[1])
+    # write out to ./out.txt for easier debugging
```
Comment (ctheune) [inline src/batou/tests/test_endtoend.py:101]: clean up on aisle 4, please

**Review by ctheune — CHANGES_REQUESTED** [review]
I like this a lot. A few minor change requested annotated. And then please squash. :)

File: src/batou/tests/test_dependencies.py (hunk 3 above)
Comment (elikoga) [reply src/batou/tests/test_dependencies.py:132]: #325

**Review by ctheune — APPROVED** [review]

**Pull request merged**